src/Psalm/Type/Atomic.php

```diff
@@ -597,6 +597,21 @@ public function replaceTemplateTypesWithArgTypes(
         // do nothing
     }
 
+    public function hasTemplate(): bool
+    {
+        if ($this instanceof TTemplateParam) {
+            return true;
+        }
+        
+        foreach ($this->getChildNodes() as $childType) {
+            if ($childType->hasTemplate()) {
```

Comment:
This subtly changes the nature of this call, and probably breaks some existing assumptions.

I think it would be better to either have a function `hasNestedTemplateParam` for this specific check, or use an instance of `TemplateTypeCollector` which does basically the same thing.

Reply:
Yeah, I was thinking earlier that this is confusing since it makes `hasTemplate` work differently from `hasArray`, `hasObject`, etc. And then I noticed NodeVisitor and wondered if that wasn't the way to go. Of course what I'd _really_ like is to figure out how to support it properly, but I'm still working my way through understanding everything.

Side note, I noticed `Union::$types` uses `'array'` for both `TArray` and `TList`, which makes [this](https://psalm.dev/r/48b13f8b7e) a bit screwy. I was just thinking about separating them but that's no good because you'd still have [this issue](https://psalm.dev/r/c23d141cfa). Is it worth opening an issue or is that something that's not worth trying to support?

Reply:
That's by design – `list` is a subtype of `array`, so `array<int, int>|list<string>` is always simplified to `array<int, int|string>`